### PR TITLE
Added the OFEP process/ contribution guide to readme.md

### LINKED ISSUE #63: Document the OFEP lifecycle
Stages:

1. Clear main readme and contribution guide
2. Issue template
3. PR based on a template
4. Set reviewers automatically based on OFEP settings
5. Once the appropriate number of approvals have been reached, post a message that the OFEP will be merged after X number of days unless someone objects
6. Automerge after the defined timeframe
7. Publish to the openfeature.dev website

### PR DESCRIPTION
**Pull request by mihirm21**
## This PR

- explains possibly the whole ofep process and acts as a contribution guide

### Related Issues

Fixes #63

### CODE REVIEW

**Comment by mihirm21** [comment]
@toddbaert @DavidPHirsch @beeme1mr

**Review by DavidPHirsch — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -1,8 +1,60 @@
-# Open Feature Enhancement Proposals
+# Open Feature Enhancement Proposals(OFEP)
 
 This repository serves as a focal point for research and experimental work.
 It also enables the creation of discussions, proposals and ideation through issues.
 
-We use OFEP: Open Feature Enhancement proposals, which comes from the lineage of [PEP](https://peps.python.org/pep-0001/) much like the Kubernetes project uses [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/README.md)
+We use OFEP: Open Feature Enhancement proposals, which comes from the lineage of [PEP](https://peps.python.org/pep-0001/) much like the Kubernetes project uses [KEP](https://github.com/OpenFeature/enhancements/blob/master/keps/README.md) and Open-Telemetry project uses [OTEP](https://github.com/open-telemetry/oteps/blob/main/README.md).
```
Comment (Kavindu-Dodan) [inline README.md:6]:
```suggestion
We use OFEP: Open Feature Enhancement proposals, which comes from the lineage of [PEP](https://peps.python.org/pep-0001/) much like the Kubernetes project uses [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/README.md) and Open-Telemetry project uses [OTEP](https://github.com/open-telemetry/oteps/blob/main/README.md).
```

**Review by Kavindu-Dodan — APPROVED** [review]

File: README.md [hunk 2]
```diff
@@ -1,8 +1,60 @@
...
\ No newline at end of file
+## Is My Thing an enhancement?
+
+We are trying to figure out the exact shape of an enhancement. Until then, here are a few rough heuristics.
+
+An enhancement is anything that:
+
+- a blog post would be written about after its release (eg. [Client-side Feature Flagging](https://openfeature.dev/blog/catering-to-the-client-side))
+- requires multiple parties/owners participating to complete (eg. Client-side Feature Flagging [Specification & SDKs])
+- will be graduating from one stage to another
```
Comment (toddbaert) [inline README.md:16]: I'm not sure about this one. Can you help me understand it?

File: README.md [hunk 3]
```diff
@@ -1,8 +1,60 @@
...
+- needs significant effort or changes OpenFeature in a significant way (eg. something that would take 10 person-weeks to implement, introduce or redesign a system component, or introduces Specification changes)
+- impacts the UX or operation of OpenFeature substantially such that engineers using OpenFeature will need retraining
+- users will notice and come to rely on
+
+It is unlikely an enhancement if it is:
+- rephrasing, grammatical fixes, typos, etc
+- bug fixes
+- refactoring code
+- adding error messages or events
+- a thing that affects only a single language or implementation
```
Comment (toddbaert) [inline README.md:26]: Strongly agree with this, nice!

File: README.md [hunk 4]
```diff
@@ -1,8 +1,60 @@
...
+We are trying to figure out the exact shape of an enhancement. Until then, here are a few rough heuristics.
+
+An enhancement is anything that:
+
+- a blog post would be written about after its release (eg. [Client-side Feature Flagging](https://openfeature.dev/blog/catering-to-the-client-side))
+- requires multiple parties/owners participating to complete (eg. Client-side Feature Flagging [Specification & SDKs])
+- will be graduating from one stage to another
+- needs significant effort or changes OpenFeature in a significant way (eg. something that would take 10 person-weeks to implement, introduce or redesign a system component, or introduces Specification changes)
+- impacts the UX or operation of OpenFeature substantially such that engineers using OpenFeature will need retraining
+- users will notice and come to rely on
```
Comment (toddbaert) [inline README.md:19]: Might be worth adding something like "impacts multiple implementations or languages" (basically the inverse of the last point below.

File: README.md [hunk 5]
```diff
@@ -1,8 +1,60 @@
...
+
+## Discussing the OFEP
+
+- As soon as the above-mentioned issue is created, potential reviewers (based on the `category of proposal` specified in the issue) would be automatically asked for review. This is done as a preventive measure to avoid long-winded and open-ended discussions in the early design phase of the proposal but the OFEP author should use their discretion here.
+- The OFEP author may choose any place for discussions but needs to be linked to the issue. The suggested ones include continuing in the same Github issue or creating a thread on Slack(https://cloud-native.slack.com/archives/C0344AANLA1) mentioning the issue. 
+- The OFEP authors are responsible for collecting community feedback on an OFEP before submitting it as a proposal for review. 
+
+## Submitting the OFEP and life-cycle
+
+- Once the idea of the proposal gets reviewed by the assigned reviewers, the OFEP author upon their discretion can then reference a Pull Request to the issue containing the proposal.
```
Comment (toddbaert) [inline README.md:47]:
```suggestion
- Once the idea of the proposal is reviewed by the assigned reviewers, the OFEP author can then reference a Pull Request to the issue containing the proposal. 
```

File: README.md [hunk 6]
```diff
@@ -1,8 +1,60 @@
...
+- Copy [`000-OFEP-template.md`](./000-OFEP-template.md) to `000-OFEP-my-title.md`, where `my-title` is a title relevant to your proposal, and `000` is the OFEP ID. Leave the number as is for now. Once a Pull Request is made, update this ID to match the next smallest available ID.
+- Fill in the template and please take care of the details as followed while creating the issue for Enhancement Proposal.
+- The initial `status` of an OFEP should be in `drafting` or `pending for review` stage.
+- An OFEP is `approved` when atleast two/three reviewers github-approve the PR but this surely depends on its nature. The OFEP is then merged.
+- If an OFEP is `rejected` or `withdrawn`, the PR is closed. Note that these OFEPs submissions are still recorded, as Github retains both the discussion and the proposal, even if the branch is later deleted.
+- If an OFEP discussion becomes long, and the OFEP then goes through a major revision, the next version of the OFEP can be posted as a new PR, which references the old PR. The old PR is then closed. This makes OFEP review easier to follow and participate in.
```
Comment (toddbaert) [inline README.md:54]:
```suggestion
- If an OFEP discussion becomes long, or the OFEP then goes through a major revision, the next version of the OFEP can be posted as a new PR, which references the old PR. The old PR is then closed. This makes OFEP review easier to follow and participate in.
```

**Review by toddbaert — APPROVED** [review]
I like this! I've left optional comments.

[This point](https://github.com/open-feature/ofep/pull/70#discussion_r1256322628) is the only one that confuses me a bit.

File: README.md (hunk 2 above)
Comment (mihirm21) [reply README.md:16]: I thought that some projects would have different packages for final and beta version, like flagd and flagd proxy.

File: README.md [hunk 7]
```diff
@@ -1,8 +1,61 @@
...
 It also enables the creation of discussions, proposals and ideation through issues.
 
-We use OFEP: Open Feature Enhancement proposals, which comes from the lineage of [PEP](https://peps.python.org/pep-0001/) much like the Kubernetes project uses [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/README.md)
+We use OFEP: Open Feature Enhancement proposals, which comes from the lineage of [PEP](https://peps.python.org/pep-0001/) much like the Kubernetes project uses [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/README.md) and Open-Telemetry project uses [OTEP](https://github.com/open-telemetry/oteps/blob/main/README.md).
 
-Please follow the [OFEP Template](000-OFEP-template.md) when creating a pull request for a new OFEP document.
\ No newline at end of file
+## Is My Thing an enhancement?
+
+We are trying to figure out the exact shape of an enhancement. Until then, here are a few rough heuristics.
```
Comment (beeme1mr) [inline README.md:10]:
```suggestion
```

I think we have a solid understanding of what requires an enhancement.

File: README.md [hunk 8]
```diff
@@ -1,8 +1,61 @@
-# Open Feature Enhancement Proposals
+# Open Feature Enhancement Proposals(OFEP)
```
Comment (beeme1mr) [inline README.md:1]:
```suggestion
# OpenFeature Enhancement Proposals(OFEP)
```

File: README.md [hunk 9]
```diff
@@ -1,8 +1,61 @@
...
 
 This repository serves as a focal point for research and experimental work.
 It also enables the creation of discussions, proposals and ideation through issues.
 
-We use OFEP: Open Feature Enhancement proposals, which comes from the lineage of [PEP](https://peps.python.org/pep-0001/) much like the Kubernetes project uses [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/README.md)
+We use OFEP: Open Feature Enhancement proposals, which comes from the lineage of [PEP](https://peps.python.org/pep-0001/) much like the Kubernetes project uses [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/README.md) and Open-Telemetry project uses [OTEP](https://github.com/open-telemetry/oteps/blob/main/README.md).
```
Comment (beeme1mr) [inline README.md:6]:
```suggestion
We use OFEP: OpenFeature Enhancement proposals, which comes from the lineage of [PEP](https://peps.python.org/pep-0001/) much like the Kubernetes project uses [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/README.md) and Open-Telemetry project uses [OTEP](https://github.com/open-telemetry/oteps/blob/main/README.md).
```

File: README.md (hunk 2 above)
Comment (beeme1mr) [reply README.md:16]: I don't think we would need an OFEP for that.

File: README.md [hunk 10]
```diff
@@ -1,8 +1,61 @@
...
+## Is My Thing an enhancement?
+
+We are trying to figure out the exact shape of an enhancement. Until then, here are a few rough heuristics.
+
+An enhancement is anything that:
+
+- a blog post would be written about after its release (eg. [Client-side Feature Flagging](https://openfeature.dev/blog/catering-to-the-client-side))
+- requires multiple parties/owners participating to complete (eg. Client-side Feature Flagging [Specification & SDKs])
+- will be graduating from one stage to another
+- needs significant effort or changes OpenFeature in a significant way (eg. something that would take 10 person-weeks to implement, introduce or redesign a system component, or introduces Specification changes)
```
Comment (beeme1mr) [inline README.md:17]:
```suggestion
- requires significant effort or modifications to OpenFeature (eg. something that would take 10 person-weeks to implement, introduce or redesign a system component, or introduces Specification changes)
```

File: README.md [hunk 11]
```diff
@@ -1,8 +1,61 @@
...
+
+### OFEP Scope
+
+While OFEPs are intended for "significant" changes, we recommend trying to keep each OFEP's scope as small as makes sense (eg. on broader scale, mentioning the category of the proposal). A general rule of thumb is that if the core functionality proposed could still provide value without a particular piece, then that piece should be removed from the proposal and used instead as an *example* (and, ideally, given its own OFEP!).
+
+## How to start with writing an OFEP? 
+
+- First, [create an issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/creating-an-issue) using the [Enhancement Proposal](https://github.com/open-feature/ofep/issues/new?assignees=beeme1mr&labels=OFEP&projects=&template=Proposal.yaml&title=%5BProposal%5D+) template.
```
Comment (beeme1mr) [inline README.md:37]:
```suggestion
- First, [create an issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/creating-an-issue) using the [Enhancement Proposal](https://github.com/open-feature/ofep/issues/new?labels=OFEP&projects=&template=Proposal.yaml&title=%5BProposal%5D+) template.
```

Let's let the assignee script do its magic.

File: README.md [hunk 12]
```diff
@@ -1,8 +1,61 @@
...
+- Fill in the template. Put care into the details: It is important to present convincing motivation, demonstrate an understanding of the design's impact, and honestly assess the drawbacks and potential alternatives.
+
+## Discussing the OFEP
+
+- As soon as the above-mentioned issue is created, potential reviewers (based on the `category of proposal` specified in the issue) would be automatically asked for review. This is done as a preventive measure to avoid long-winded and open-ended discussions in the early design phase of the proposal but the OFEP author should use their discretion here.
+- The OFEP author may choose any place for discussions but needs to be linked to the issue. The suggested ones include continuing in the same Github issue or creating a thread on Slack(https://cloud-native.slack.com/archives/C0344AANLA1) mentioning the issue.
```
Comment (beeme1mr) [inline README.md:43]:
```suggestion
- The OFEP author may choose any place for discussions but needs to be linked to the issue. The suggested ones include continuing in the same GitHub issue or creating a thread on Slack(https://cloud-native.slack.com/archives/C0344AANLA1) mentioning the issue. 
```

File: README.md [hunk 13]
```diff
@@ -1,8 +1,61 @@
...
+
+- Once the idea of the proposal is reviewed by the assigned reviewers, the OFEP author can then reference a Pull Request to the issue containing the proposal. 
+- For adding the OFEP as a Pull Request, first, [fork](https://help.github.com/en/articles/fork-a-repo) this [repo](https://github.com/open-feature/ofep).
+- Copy [`000-OFEP-template.md`](./000-OFEP-template.md) to `000-OFEP-my-title.md`, where `my-title` is a title relevant to your proposal, and `000` is the OFEP ID. Leave the number as is for now. Once a Pull Request is made, update this ID to match the next smallest available ID.
+- Fill in the template and please take care of the details as followed while creating the issue for Enhancement Proposal.
+- The initial `status` of an OFEP should be in `drafting` or `pending for review` stage.
+- An OFEP is `approved` when atleast two/three reviewers github-approve the PR but this surely depends on its nature. The OFEP is then merged.
+- If an OFEP is `rejected` or `withdrawn`, the PR is closed. Note that these OFEPs submissions are still recorded, as Github retains both the discussion and the proposal, even if the branch is later deleted.
```
Comment (beeme1mr) [inline README.md:54]:
```suggestion
- If an OFEP is `rejected` or `withdrawn`, the PR is closed. Note that these OFEPs submissions are still recorded, as GitHub retains both the discussion and the proposal, even if the branch is later deleted.
```

File: README.md [hunk 14]
```diff
@@ -1,8 +1,61 @@
...
+- If an OFEP discussion becomes long, or the OFEP then goes through a major revision, the next version of the OFEP can be posted as a new PR, which references the old PR. The old PR is then closed. This makes OFEP review easier to follow and participate in.
+
+## Implementing the OFEP
+
+Some accepted OFEPs represent vital features that need to be implemented right away. Other accepted OFEPs can represent features that can wait until some arbitrary developer feels like doing the work. Every accepted OFEP has an associated issue tracking its implementation in the repository specific to it.
```
Comment (beeme1mr) [inline README.md:59]:
```suggestion
Some accepted OFEPs represent vital features that need to be implemented right away. Other accepted OFEPs can represent features that can wait until a community member decides to implement the functionality. Every accepted OFEP has an associated issue tracking its implementation in the specific repository.
```

File: README.md [hunk 15]
```diff
@@ -1,8 +1,61 @@
...
 
 This repository serves as a focal point for research and experimental work.
 It also enables the creation of discussions, proposals and ideation through issues.
 
-We use OFEP: Open Feature Enhancement proposals, which comes from the lineage of [PEP](https://peps.python.org/pep-0001/) much like the Kubernetes project uses [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/README.md)
+We use OFEP: Open Feature Enhancement proposals, which comes from the lineage of [PEP](https://peps.python.org/pep-0001/) much like the Kubernetes project uses [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/README.md) and Open-Telemetry project uses [OTEP](https://github.com/open-telemetry/oteps/blob/main/README.md).
 
-Please follow the [OFEP Template](000-OFEP-template.md) when creating a pull request for a new OFEP document.
\ No newline at end of file
+## Is My Thing an enhancement?
```
Comment (beeme1mr) [inline README.md:8]:
```suggestion
## Should I create an OFEP?
```

File: README.md [hunk 16]
```diff
@@ -1,8 +1,61 @@
...
+
+We are trying to figure out the exact shape of an enhancement. Until then, here are a few rough heuristics.
+
+An enhancement is anything that:
```
Comment (beeme1mr) [inline README.md:12]:
```suggestion
You should create an OFEP for anything that:
```

**Review by beeme1mr — COMMENTED** [review]
Looks good! I think we're getting close. Thanks for putting this together.

**Review by beeme1mr — APPROVED** [review]

**Pull request merged**